File: 다이내믹 프로그래밍/11053.가장긴증가하는부분수열.js
//[실버2] 11053.가장긴증가하는부분수열 https://www.acmicpc.net/problem/11053

import { createRequire } from "module";
const require = createRequire(import.meta.url);

const fs = require('fs');
let input = fs.readFileSync('dev/stdin', 'utf-8').trim().split('\n');

let n = +input[0]
let A = input[1].split(' ').map((v)=>+v);

function solution(n, A){

    let dp = Array(n).fill(1);

    for(let i=1; i<n;i++){
        for(let j=i;j>=0;j--){
            if(A[i] > A[j]){
                dp[i] = Math.max(dp[i], dp[j]+1);
            }
        }
    }

    console.log(Math.max(...dp));

}

solution(n, A);

File: 다이내믹 프로그래밍/11053.가장긴증가하는부분수열.py
#[실버2] 11053.가장긴증가하는부분수열 https://www.acmicpc.net/problem/11053

import sys

n = int(sys.stdin.readline())
arr = list(map(int, sys.stdin.readline().split()))

def solution(n, arr):
    dp = [1] * n
    
    for i in range(1, n):
        for j in range(i-1, -1, -1):
            if arr[j] < arr[i]:
                dp[i] = max(dp[j] + 1, dp[i])

    return max(dp)

print(solution(n, arr))